Plutarch/Api/V1.hs

```diff
@@ -0,0 +1,572 @@
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Api.V1 (
+  -- * V1 Specific types
+  PScriptContext (..),
+  PTxInfo (..),
+
+  -- * General types, compatible with V1 and V2
+  PScriptPurpose (..),
+
+  -- ** Script
+  PDatum (..),
+  PRedeemer (..),
+  PDatumHash (..),
+  PRedeemerHash (..),
+  PValidatorHash (..),
+  PStakeValidatorHash (..),
+
+  -- ** Value
+  PValue (..),
+  PCurrencySymbol (..),
+  PTokenName (..),
+
+  -- ** Crypto
+  PPubKeyHash (..),
+  PPubKey (..),
+  PSignature (..),
+
+  -- ** Time
+  PPOSIXTime (..),
+  type PPOSIXTimeRange,
+
+  -- ** Interval
+  PInterval (..),
+  PLowerBound (..),
+  PUpperBound (..),
+  PExtended (..),
+  type PClosure,
+
+  -- ** Address
+  PAddress (..),
+  PCredential (..),
+  PStakingCredential (..),
+
+  -- ** Tx
+  PTxInInfo (..),
+  PTxOut (..),
+  PTxOutRef (..),
+  PTxId (..),
+  PDCert (..),
+
+  -- ** AssocMap
+  PMap (..),
+
+  -- ** Others
+  PMaybe (..),
+  PEither (..),
+) where
+
+--------------------------------------------------------------------------------
+
+import Plutarch (PMatch)
+import Plutarch.Bool (PBool)
+import Plutarch.Builtin (PAsData, PBuiltinList, PData, PIsData, type PBuiltinMap)
+import Plutarch.ByteString (PByteString)
+import Plutarch.DataRepr (
+  DataReprHandlers (DRHCons, DRHNil),
+  DerivePConstantViaData (DerivePConstantViaData),
+  PDataList,
+  PIsDataRepr,
+  PIsDataReprInstances (PIsDataReprInstances),
+  PIsDataReprRepr,
+  pmatchDataRepr,
+  pmatchRepr,
+ )
+import Plutarch.Integer (PInteger, PIntegral)
+import Plutarch.Lift (DerivePConstantViaNewtype (DerivePConstantViaNewtype), PConstant, PLifted, PUnsafeLiftDecl)
+
+-- ctor in-scope for deriving
+import Plutarch.Prelude
+import qualified Plutus.V1.Ledger.Api as Plutus
+import qualified PlutusTx.Builtins.Internal as PT
+
+--------------------------------------------------------------------------------
+
+type PTuple = PDataList
+
+---------- V1 Specific types, Incompatible with V2
+
+newtype PTxInfo (s :: S)
```

Comment:
I just noticed this. I think we should keep these types as `data` rather than `newtype`. Similar to their ledger api variants.

It lessens the cognitive load when I can glance at a data type to know whether or not its runtime representation (i.e runtime as in Plutus core, not haskell runtime) is just its inner field. In the case of `TxInfo`, `ScriptContext` and similar - that is not the case, their constructors does indeed persist during runtime. It'd be helpful to *semantically* reflect that in Plutarch synonyms IMO

CC @L-as

Reply:
Well, in the case of `TxInfo`, it *is* just its inner field.

Reply:
It's not though. `data TxInfo = TxInfo ...`. It's the constructor + the fields in the plutus core. Whereas the `newtype` types conventionally don't keep their constructors.

This distinction has helped me hover through plutus ledger types and understand what their plutus core representations are like. I've used the thumb rule: If it's a `newtype`, there is no extra `Constr` step - it's just the data representation of its singular field. If it's a `data`, there is a `Constr` (unless it's an integer, bytestring, list, or map) + the field(s). It'd be nice to keep this distinction in Plutarch synonyms - though it is semantically redundant.

Reply:
That's a good point and the distinction has actually tripped me up somewhat, too

Reply:
It's a newtype of `PDataList`. `PDataList` itself is a `Constr`.

Reply:
Sure, but `PDataList` is very much a Plutarch concept. Keeping the plutus `data` types as `data`, and `newtype` types as `newtype` makes it immediately obvious IMO. I don't think it costs us anything either.

Reply:
It does, it makes it impossible to require that the contents match something.

Reply:
(Through Coercible)

Reply:
> It does, it makes it impossible to require that the contents match something.
> 
> (Through Coercible)

It can be done using `generics-sop` (something like `Code (a s) ~ '[ '[ PDataList XS ] ]`, but it will require that `SI` type I mentioned in Discord.

Reply:
> It does, it makes it impossible to require that the contents match something.
> 
> (Through Coercible)

Ugh, that's unfortunate.